assets/php/class-wgpb-block-grid-base.php

```diff
@@ -240,64 +254,51 @@ protected function get_container_classes() {
 	 * @return string Rendered product output.
 	 */
 	public function render_product( $id ) {
-		global $product;
 		$product = wc_get_product( $id );
+
 		if ( ! $product ) {
 			return '';
 		}
 
-		$title = sprintf(
-			'<div class="wc-block-grid__product-title">%s</div>',
-			wp_kses_post( $product->get_title() )
-		);
-
-		$rating_str = $this->get_rating( $product );
-
-		$price_str = sprintf(
-			'<div class="wc-block-grid__product-price price">%s</div>',
-			$product->get_price_html()
-		);
-
-		ob_start();
-		woocommerce_show_product_sale_flash();
-		$on_sale_flag = ob_get_contents();
-		ob_end_clean();
-
-		$button_str = sprintf(
-			'<div class="wc-block-grid__product-add-to-cart">%s</div>',
-			$this->get_add_to_cart( $product )
+		// Get escaped data to insert into template.
+		$data = (object) array(
+			'permalink'  => $product->get_permalink(),
+			'image'      => $product->get_image( 'woocommerce_thumbnail' ),
+			'title'      => $product->get_title(),
+			'rating'     => $this->get_rating( $product ),
+			'sale_badge' => $this->get_sale_badge( $product ),
+			'price'      => $product->get_price_html(),
+			'button'     => $this->get_add_to_cart( $product ),
 		);
 
-		$content = sprintf(
-			'<a href="%s" class="wc-block-grid__product-link">',
-			$product->get_permalink()
-		);
-
-		$content .= sprintf(
-			'<div class="wc-block-grid__product-image">%s</div>',
-			woocommerce_get_product_thumbnail()
-		);
+		return "
+		<li class=\"wc-block-grid__product product\">
+			<a href=\"{$data->permalink}\" class=\"wc-block-grid__product-link\">
+				<div class=\"wc-block-grid__product-image\">{$data->image}</div>
+				<div class=\"wc-block-grid__product-title\">{$data->title}</div>
```

Comment:
Same comment as above – we should check the `contentVisibility` attribute before rendering these.